Sming/Wiring/WString.cpp

```diff
@@ -598,85 +607,91 @@ unsigned int String::getBytes(unsigned char *buf, unsigned int bufsize, unsigned
-  return temp - buffer;
-}
-
-int String::indexOf(const String &s2) const
-{
-  return indexOf(s2, 0);
+  return static_cast<const char*>(temp) - buf;
 }
 
-int String::indexOf(const String &s2, unsigned int fromIndex) const
+int String::indexOf(const char* s2_buf, size_t s2_len, size_t fromIndex) const
 {
+  auto len = length();
   if (fromIndex >= len) return -1;
-  auto found = (const char*)memmem(buffer + fromIndex, len - fromIndex, s2.buffer, s2.len);
+  auto buf = cbuffer();
+  auto found = memmem(buf + fromIndex, len - fromIndex, s2_buf, s2_len);
   if (found == nullptr) return -1;
-  return found - buffer;
+  return static_cast<const char*>(found) - buf;
 }
 
 int String::lastIndexOf(char theChar) const
 {
-  return lastIndexOf(theChar, len - 1);
+	auto len = length();
+	if(len == 0) {
+		return -1;
+	}
+	auto buf = cbuffer();
+	auto found = memrchr(buf, theChar, len);
+	return found ? (static_cast<const char*>(found) - buf) : -1;
 }
 
-int String::lastIndexOf(char ch, unsigned int fromIndex) const
+int String::lastIndexOf(char ch, size_t fromIndex) const
 {
-  if (fromIndex >= len) return -1;
-  char tempchar = buffer[fromIndex + 1];
-  buffer[fromIndex + 1] = '\0';
-  char* temp = strrchr(buffer, ch);
-  buffer[fromIndex + 1] = tempchar;
-  if (temp == nullptr) return -1;
-  return temp - buffer;
+	auto len = length();
+	if(len == 0) {
+		return -1;
+	}
+	if(fromIndex < len) {
+		len = fromIndex + 1;
+	}
+	auto buf = cbuffer();
+	auto found = memrchr(buf, ch, len);
+	return found ? (static_cast<const char*>(found) - buf) : -1;
 }
 
 int String::lastIndexOf(const String &s2) const
 {
-  return lastIndexOf(s2, len - s2.len);
+  return lastIndexOf(s2.cbuffer(), s2.length(), length() - s2.length());
+}
+
+int String::lastIndexOf(const String &s2, size_t fromIndex) const
+{
+	return lastIndexOf(s2.cbuffer(), s2.length(), fromIndex);
 }
 
-int String::lastIndexOf(const String &s2, unsigned int fromIndex) const
+int String::lastIndexOf(const char* s2_buf, size_t s2_len, size_t fromIndex) const
 {
-  if (s2.len == 0 || len == 0 || s2.len > len) return -1;
+  auto len = length();
+  if (s2_len == 0 || len == 0 || s2_len > len) return -1;
+  auto buf = cbuffer();
   if (fromIndex >= len) fromIndex = len - 1;
   int found = -1;
-  for (char *p = buffer; p <= buffer + fromIndex; p++)
+  for (auto p = buf; p <= buf + fromIndex; p++)
   {
-    p = (char*)memmem(p, buffer + len - p, s2.buffer, s2.len);
+    p = static_cast<const char*>(memmem(p, buf + len - p, s2_buf, s2_len));
     if (!p) break;
-    if (p <= buffer + fromIndex) found = p - buffer;
+    if (p <= buf + fromIndex) found = p - buf;
   }
   return found;
 }
 
-String String::substring(unsigned int left, unsigned int right) const
+String String::substring(size_t left, size_t right) const
 {
-  if (!buffer) return nullptr;
+  if (isNull()) return nullptr;
 
   if (left > right)
   {
-    unsigned int temp = right;
+    size_t temp = right;
     right = left;
     left = temp;
   }
```

Comment:
This check is present in both AVR and Esp8266 Arduino cores. Maybe there's a historical reason for it but in my view the method should fail if the parameters have been interchanged. Here's the doc. page:
https://www.arduino.cc/reference/en/language/variables/data-types/string/functions/substring/

Reply:
@slaff any thoughts on this one? Should we change it?

Reply:
Yes, change it please. But put this change as a separate PR. If we break some Arduino libraries it should be easy to revert it.